addon/lib/StudyTelemetryCollector.jsm

```diff
@@ -21,15 +22,266 @@ const BASERESOURCE = "esper-pioneer-shield-study";
+    attributes.scalar_parent_browser_engagement_max_concurrent_tab_count = payload.processes.parent.scalars["browser.engagement.max_concurrent_tab_count"];
+    attributes.scalar_parent_browser_engagement_tab_open_event_count = payload.processes.parent.scalars["browser.engagement.tab_open_event_count"];
+    attributes.scalar_parent_browser_engagement_window_open_event_count = payload.processes.parent.scalars["browser.engagement.window_open_event_count"];
+    attributes.scalar_parent_browser_engagement_unique_domains_count = payload.processes.parent.scalars["browser.engagement.unique_domains_count"];
+    attributes.scalar_parent_browser_engagement_total_uri_count = payload.processes.parent.scalars["browser.engagement.total_uri_count"];
+    attributes.scalar_parent_browser_engagement_unfiltered_uri_count = payload.processes.parent.scalars["browser.engagement.unfiltered_uri_count"];
+    attributes.scalar_parent_browser_engagement_navigation_about_newtab = payload.processes.parent.scalars["browser.engagement.navigation_about_newtab"];
+    attributes.scalar_parent_browser_engagement_navigation_contextmenu = payload.processes.parent.scalars["browser.engagement.navigation_contextmenu"];
+    attributes.scalar_parent_browser_engagement_navigation_searchbar = payload.processes.parent.scalars["browser.engagement.navigation_searchbar"];
+    attributes.scalar_parent_browser_engagement_navigation_urlbar = payload.processes.parent.scalars["browser.engagement.navigation_urlbar"];
+
+    return attributes;
+
   }
+
+  /**
+   * Collects the places-db-related telemetry by querying the database directly rather relying on telemetry since, due to performance reasons,
+   * regular telemetry only collects this telemetry occasionally.
+   *
+   * An earlier attempt to gather the places-db-related telemetry was to await PlacesDBUtils.telemetry() and then fetch the populated histograms
+   * but PlacesDBUtils.telemetry() does not return a promise, it just updates the current telemetry session with additional histograms when db queries are ready
+   * and I found no reliant way to ensure these histograms being available in the current telemetry session.
+   *
+   * @returns {Promise}
+   */
+  static async collectPlacesDbBasedAttributes() {
+
+    return new Promise((resolve, reject) => {
+
+      StudyTelemetryCollector.queryPlacesDbTelemetry().then((placesDbTelemetryResults) => {
+
+        resolve({
+          places_pages_count: placesDbTelemetryResults[0][1],
+          places_bookmarks_count: placesDbTelemetryResults[1][1],
+        });
+
+      }).catch(ex => reject(ex));
+
+    });
+
+  }
+
+  /**
+   * Based on code in firefox/toolkit/components/places/PlacesDBUtils.jsm
+   *
+   * @returns {Promise.<A|Promise.<*[]>|Promise.<Array.<T>>>}
+   */
+  static async queryPlacesDbTelemetry() {
+
+    const probes = [
+      {
+        histogram: "PLACES_PAGES_COUNT",
```

Comment:
Is this manually filling the data into the histogram itself, or it's just mentioning "histogram" due to how the PlacesDB API works?

Reply:
It only mentions histogram since the code it is based on mentioned it. It does not feed any data into histograms. This whole manual querying was implemented since [I couldn't find a decent way to ensure that these histograms actually contained data](https://github.com/motin/esper-pioneer-shield-study/pull/2/files#diff-345bfc78852ae1d161ce3aedab81f9d6R208)

Reply:
Ah, cool, ok then :)